Adds a MeshCache to _msgsToAttachedBody to prevent duplicate attached collision objects being stored in FCLShapeCache

rr-mark opened the pull request:
### Description

Fixes https://github.com/ros-planning/moveit/issues/3515

Previously, when planning a sequence motion, the attached collision objects would be converted to separate shapes::Mesh* for each waypoint along the path. Then each instance would be come a separate fcl::BVHModel instance in the FCLShapeCache, leading to large computational costs and memory usage when using large attached collision objects.

This PR adds caching of shapes constructed from meshes to _msgsToAttachedBody in conversions.cpp. This means that the shapes::Mesh* pointers passed into the FCLShapeCache are correctly aliased, and fcl::BVHModels do not need to be regenerated.

The mesh cache has an (approximate) maximum memory usage, and evicts cache entries using a last-used-first-out strategy when this memory usage is exceeded.

### Checklist
- [✓] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/) N/A (I think?)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes N/A (no API changes)
- [✓] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI N/A (no GUI changes)
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

rr-mark commented:
> I really recommend you to install pre-commit, saves some round trips with the ci. https://moveit.ros.org/documentation/contributing/code/

Yeah, sorry, my editor had the wrong clang-format loaded and kept undoing my changes. I think it's fixed now.

simonschmeisser commented:
Very nice PR, thanks!

> Some open questions I have on this:
> 
>     * What values should be used for min_size_to_cache and max_cache size? The 1Mb and 1Gb values are just guesses. I feel like these should probably be user-configurable, although I don't know how that process works here.

Did you measure the overhead? Is there a size where it does not help to use the cache? I would suspect the turn-even << 1MB?
> 
>     * Is there a better way to hash a mesh than converting it to string and hashing the string?

Maybe calculating hashes directly on `mesh.triangles.data()` and `mesh.vertices.data()`? 
> 
>     * The MeshCacheControlBlock currently stores both the ShapeConstPtr and the std::string mesh, which makes it twice as large as it maybe needs to be. Any smart ideas on how to guarantee meshes are correct without storing both?

So this basically boils down to the question if hash collisions are likely (enough) to warrant the storage costs. Maybe storing something like number of vertices and/or edges would be sufficient?

rr-mark commented:
> > * What values should be used for min_size_to_cache and max_cache size? The 1Mb and 1Gb values are just guesses. I feel like these should probably be user-configurable, although I don't know how that process works here.
> 
> Did you measure the overhead? Is there a size where it does not help to use the cache? I would suspect the turn-even << 1MB?

I haven't measured any overhead, and I'm not sure if my use cases are representative of wider use. I think the use case for the minimum size is if someone has a few large collision objects and many small ones, then the idea is to not cache the small objects to avoid kicking the large objects out of the cache. But I guess that use case is relatively specific, so the min size should default to zero?

For the maximum cache size, I guess it depends on how much RAM the user's computer has.

> > * Is there a better way to hash a mesh than converting it to string and hashing the string?
> 
> Maybe calculating hashes directly on `mesh.triangles.data()` and `mesh.vertices.data()`?

Makes sense. I'll try that and see how it goes.

> > * The MeshCacheControlBlock currently stores both the ShapeConstPtr and the std::string mesh, which makes it twice as large as it maybe needs to be. Any smart ideas on how to guarantee meshes are correct without storing both?
> 
> So this basically boils down to the question if hash collisions are likely (enough) to warrant the storage costs. Maybe storing something like number of vertices and/or edges would be sufficient?

I was assuming hash collisions were very unlikely, but that the cache returning the wrong shape would be critically bad.

rr-mark commented:
> > > * What values should be used for min_size_to_cache and max_cache size? The 1Mb and 1Gb values are just guesses. I feel like these should probably be user-configurable, although I don't know how that process works here.
> > 
> > 
> > Did you measure the overhead? Is there a size where it does not help to use the cache? I would suspect the turn-even << 1MB?
> 
> I haven't measured any overhead, and I'm not sure if my use cases are representative of wider use. I think the use case for the minimum size is if someone has a few large collision objects and many small ones, then the idea is to not cache the small objects to avoid kicking the large objects out of the cache. But I guess that use case is relatively specific, so the min size should default to zero?
> 
> For the maximum cache size, I guess it depends on how much RAM the user's computer has.
> 
> > > * Is there a better way to hash a mesh than converting it to string and hashing the string?
> > 
> > 
> > Maybe calculating hashes directly on `mesh.triangles.data()` and `mesh.vertices.data()`?
> 
> Makes sense. I'll try that and see how it goes.
> 
> > > * The MeshCacheControlBlock currently stores both the ShapeConstPtr and the std::string mesh, which makes it twice as large as it maybe needs to be. Any smart ideas on how to guarantee meshes are correct without storing both?
> > 
> > 
> > So this basically boils down to the question if hash collisions are likely (enough) to warrant the storage costs. Maybe storing something like number of vertices and/or edges would be sufficient?
> 
> I was assuming hash collisions were very unlikely, but that the cache returning the wrong shape would be critically bad.

I've changed the default `min_size_to_cache` to 0, and added a TODO to consider exposing these parameters to the user (I don't think I know enough about how user-defined parameters in moveit work to do a good job of this myself).

I looked into changing the hash to use the raw message type rather than the string representation, but I think this is just going to add a lot of code complexity for little gain.

rr-mark commented:
It looks like there's a problem using this PR with OMPL:

When two attached collision objects in the scene have the same mesh (and so their shape::Mesh* pointers are aliased by the MeshCache in this PR), the OMPL planner hangs forever on `ompl_parallel_plan_.solve(...)`.

Before I investigate further, I'd like to know if either:
 - my approach in this PR is right, and this is a bug in OMPL (or the OMPL wrapper in Moveit), which I should investigate
 - OR the aliasing of meshes in this PR is not desirable behaviour, and I should try some other method of de-duplicating the entries in the FCLShapeCache
 - OR some other course of action would be better

rr-mark commented:
@rhaschke could I get your thoughts on this PR and the issue with OMPL when you have the time? I'm keen to get this ready to merge, but not sure how to proceed.

simonschmeisser commented:
This is a bit surprising as I couldn't spot any obvious new locks/mutexes and the cache is thread-local.

Could you provide a backtrace from when it hangs?

```
echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope
gdb --pid=$(pidof moveit_core) -ex 'set logging on' -ex 'set pagination off' -ex 'thread apply all bt full' -ex detach -ex quit
```

rr-mark commented:
> Could you provide a backtrace from when it hangs?
Sure, I'll give that a go.

> This is a bit surprising as I couldn't spot any obvious new locks/mutexes and the cache is thread-local.
I don't think it's my code that's hanging, but rather than OMPL is hanging as a result of the aliased meshes.

rr-mark commented:
I am struggling to attach gdb to the move_group in our system due to our docker and dependency management setup. I have passed this over to @rr-tom-noble who knows more about these systems. We should have a backtrace for you soon.

While trying to attach gdb, I did notice that it is not a full hang. OMPL just goes a lot slower than it should. I'm wondering if it is some kind of cache thrashing on the OMPL end. Hopefully we'll know more shortly.

rr-mark commented:
[gdb.txt](https://github.com/ros-planning/moveit/files/13416783/gdb.txt)

rr-mark commented:
I wonder if this is the wrong solution to the problem, and if something a bit more targeted to FCL would be more appropriate?

If I can get FCL to hash the meshes in its cache, I could change that cache to identify identical meshes which don't share a pointer, and change that cache to look more like the cache I have implemented here.

That would hopefully solve the problem without so many possible downstream consequences, and not introduce a second cache where a second cache isn't really needed.

I'll probably look into doing this sometime next week.

simonschmeisser commented:
> When two attached collision objects in the scene have the same mesh (and so their shape::Mesh* pointers are aliased by the MeshCache in this PR), the OMPL planner hangs forever on `ompl_parallel_plan_.solve(...)`.

I missed this particular point (**two**) until now. The cache used in the FCL collision checker plugin (not actually part of FCL but MoveIt) assumes there's at most one attached and one non-attached instance of a certain mesh. Your use-case breaks this assumption.

https://github.com/ros-planning/moveit/blob/d8e0a72e9a69202521aad251a9fa8e2696d8934e/moveit_core/collision_detection_fcl/src/collision_common.cpp#L370

https://github.com/ros-planning/moveit/blob/d8e0a72e9a69202521aad251a9fa8e2696d8934e/moveit_core/collision_detection_fcl/src/collision_common.cpp#L695

Not sure how to fix/extend this actually ...

rr-mark commented:
Since this PR introduces downstream problems, and the performance issues appear to be fixed by https://github.com/moveit/moveit/pull/3590, I'm going to close it.
